### PR TITLE
Add recurrence rule evaluation simplified

### PR DESCRIPTION
**Pull request by rnantes**
This PR adds support for evaluating if a date satisfies a `RecurrenceRule` and for finding the next available date that satisfies the `RecurrenceRule` after a given date

### CODE REVIEW

File: Sources/Jobs/Scheduler/DateExtensions.swift [hunk 1]
```diff
@@ -112,5 +112,373 @@ internal extension Calendar {
...
+    ///
+    /// - Parameter timeZone: The TimeZone to base the date off of (defaults to current timeZone)
+    /// - Returns: The yearForWeekOfYear of the datesss
+    internal func yearForWeekOfYear(atTimeZone timeZone: TimeZone? = nil) -> Int? {
+        return calendar(atTimeZone: timeZone).component(.yearForWeekOfYear, from: self)
+    }
+
+    /// finds the number of weeks the year (52 or 53)
+    internal func weeksInYear() -> Int? {
+        // return NSCalendar.current.range(of: .weekOfYear, in: .yearForWeekOfYear, for: self)?.count
```
[the remark this answers is not in the document]
Comment (rnantes) [reply Sources/Jobs/Scheduler/DateExtensions.swift]: Removed this line.

File: Sources/Jobs/Scheduler/RecurrenceRule.swift [hunk 2]
```diff
@@ -114,3 +114,298 @@ internal struct RecurrenceRule {
...
+        }
+    }
+
+    /// Finds the next date from the starting date that satisfies the rule
+    ///
+    /// - Warning: The search is exhausted after the year 3000
+    ///
+    /// - Parameter currentDate: The starting date, a date after this will be searched for
+    /// - Returns: The next date that satisfies the rule
+    public func resolveNextDateThatSatisfiesRule(currentDate: Date) throws -> Date {
```
[the remark this answers is not in the document]
Comment (rnantes) [reply Sources/Jobs/Scheduler/RecurrenceRule.swift]: Changed to internal

File: Sources/Jobs/Scheduler/ScheduleBuilder.swift [hunk 3]
```diff
@@ -280,50 +288,99 @@ public final class ScheduleBuilder {
...
         /// The minute to run the job at
         /// - Parameter minute: A `Minute` to run the job at
         public func at(_ minute: Minute) {
             self.builder.minute = minute
         }
     }
-    
+
+    /// returns the next date that satisfies the schedule
+    public func resolveNextDateThatSatisifiesSchedule(date: Date) throws -> Date {
```
[the remark this answers is not in the document]
Comment (rnantes) [reply Sources/Jobs/Scheduler/ScheduleBuilder.swift]: Changed to internal

**Comment by mcdappdev** [comment]
@tanner0101 friendly ping on this :)

**Comment by mcdappdev** [comment]
@rnantes Great work! We're going to go ahead and merge this now and move forward with the feature. Thanks for all of your hard work on developing this, it wouldn't have happened without you.

**Comment by penny-coin** [comment]
Hey @rnantes, you just merged a pull request, have a coin! 

You now have 2 coins.

**Pull request merged**